tools/base/checker.py

```diff
@@ -0,0 +1,286 @@
+        """The "path" - usually Envoy src dir. This is used for finding configs for the tooling and should be a dir"""
+        try:
+            path = self.args.path or self.args.paths[0]
+        except IndexError:
+            raise self.parser.error(
+                "Missing path: `path` must be set either as an arg or with --path")
+        if not os.path.isdir(path):
+            raise self.parser.error(
+                "Incorrect path: `path` must be a directory, set either as first arg or with --path"
+            )
+        return path
+
+    @property
+    def paths(self) -> list:
+        """List of paths to apply checks to"""
+        return self.args.paths or [self.path]
+
+    @property
+    def recurse(self) -> bool:
+        """Flag to determine whether to apply checks recursively (where appropriate)"""
+        return self.args.recurse
+
+    @property
+    def show_summary(self) -> bool:
+        """Show a summary at the end or not"""
+        return bool(self.args.summary or self.error_count or self.warning_count)
+
+    @property
+    def status(self) -> dict:
+        """Dictionary showing current success/warnings/errors"""
+        return dict(
+            success=self.success_count,
+            errors=self.error_count,
+            warnings=self.warning_count,
+            failed=self.failed,
+            warned=self.warned,
+            succeeded=self.succeeded)
+
+    @property
+    def succeeded(self) -> dict:
+        """Dictionary of successful checks grouped by check type"""
+        return dict((k, (len(v))) for k, v in self.success.items())
+
+    @property
+    def success_count(self) -> int:
+        """Current count of successful checks"""
+        return sum(len(e) for e in self.success.values())
+
+    @cached_property
+    def summary(self) -> bool:
+        """Instance of the checker's summary class"""
+        return self.summary_class(self)
+
+    @property
+    def summary_class(self):
+        """Checker's summary class"""
+        return CheckerSummary
+
+    @property
+    def warned(self) -> dict:
+        """Dictionary of warned checks grouped by check type"""
+        return dict((k, (len(v))) for k, v in self.warnings.items())
+
+    @property
+    def warning_count(self) -> int:
+        """Current count of warned checks"""
+        return sum(len(e) for e in self.warnings.values())
+
+    def add_arguments(self, parser: argparse.ArgumentParser) -> None:
+        """Add arguments to the arg parser"""
+        parser.add_argument(
+            "--fix", action="store_true", default=False, help="Attempt to fix in place")
+        parser.add_argument(
+            "--diff",
+            action="store_true",
+            default=False,
+            help="Display a diff in the console where available")
+        parser.add_argument(
+            "--recurse",
+            "-r",
+            choices=["yes", "no"],
+            default="yes",
+            help="Recurse path or paths directories")
```

Comment:
I'm hoping we have some generic way of ensuring we only apply to checked-in git directories (and avoiding things like .git)..

Reply:
this is kinda tool specific.

i always prefer to use the git tree where i can and avoid random local tmp dirs getting linted etc

im now thinking that this should possibly not be in the base class.

i can shift it to the class where its actually used

Reply:
OK, SG, just flagging this as an common problem in tooling and worth future consideration.